File: TransporteInterdepartamental.java
/**
 * Clase principal que representa una aplicación para modelar rutas
 * de transporte interdepartamental en Guatemala usando un grafo dirigido.
 *
 * Utiliza la clase {@link Graph} para representar ciudades como vértices,
 * rutas como aristas con pesos (distancias en kilómetros) y el algoritmo
 * de Floyd-Warshall para calcular distancias mínimas entre todas las ciudades.
 *
 * Funcionalidades:
 * <ul>
 *   <li>Crear un grafo con 5 ciudades</li>
 *   <li>Agregar conexiones con distancias entre las ciudades</li>
 *   <li>Imprimir la matriz de adyacencia</li>
 *   <li>Aplicar el algoritmo de Floyd-Warshall</li>
 *   <li>Determinar el centro del grafo (ciudad con menor excentricidad)</li>
 *   <li>Mostrar las rutas más eficientes desde Ciudad de Guatemala</li>
 * </ul>
 * 
 * Las ciudades representadas son:
 * <ul>
 *   <li>0 - Ciudad de Guatemala</li>
 *   <li>1 - Zacapa</li>
 *   <li>2 - Chiquimula</li>
 *   <li>3 - Quetzaltenango</li>
 *   <li>4 - Cobán</li>
 * </ul>
 * 
 * @author Diana
 */
public class TransporteInterdepartamental {

    /**
     * Método principal que ejecuta la lógica del programa.
     * @param args Argumentos de línea de comandos (no utilizados)
     */
    public static void main(String[] args) {
        Graph graph = new Graph(5);
        
        graph.addVertex(0, "Ciudad de Guatemala");
        graph.addVertex(1, "Zacapa");
        graph.addVertex(2, "Chiquimula");
        graph.addVertex(3, "Quetzaltenango");
        graph.addVertex(4, "Cobán");
        
        graph.addEdge(0, 1, 3);
        graph.addEdge(0, 3, 7);  
        graph.addEdge(1, 2, 1);
        graph.addEdge(1, 4, 8);
        graph.addEdge(2, 3, 2);
        graph.addEdge(3, 4, 3);
        graph.addEdge(4, 0, 4);
        
        System.out.println("Matriz de adyacencia inicial:");
        graph.printMatrix();
        
        int[][] shortestPaths = graph.floydWarshall();
    
        System.out.println("\nMatriz de distancias mínimas (resultado de Floyd-Warshall):");
        for (int i = 0; i < graph.getSize(); i++) {
            for (int j = 0; j < graph.getSize(); j++) {
                if (shortestPaths[i][j] == Integer.MAX_VALUE) {
                    System.out.print("∞\t");
                } else {
                    System.out.print(shortestPaths[i][j] + "\t");
                }
            }
            System.out.println();
        }

        int center = graph.findGraphCenter();
        System.out.println("\nEl centro del grafo es: " + graph.getVertices()[center]);
        
        System.out.println("\nRutas más eficientes desde Ciudad de Guatemala:");
        for (int i = 0; i < graph.getSize(); i++) {
            if (i != 0) { 
                System.out.print("Hacia " + graph.getVertices()[i] + ": ");
                if (shortestPaths[0][i] == Integer.MAX_VALUE) {
                    System.out.println("No hay ruta disponible");
                } else {
                    System.out.println(shortestPaths[0][i] + " km");
                }
            }
        }
    }
}

File: transporte_interdepartamental.py
import networkx as nx
import numpy as np

def main():
    G = nx.DiGraph()
    
    G.add_node(0, name="Ciudad de Guatemala")
    G.add_node(1, name="Zacapa")
    G.add_node(2, name="Chiquimula")
    G.add_node(3, name="Quetzaltenango")
    G.add_node(4, name="Cobán")
    
    G.add_edge(0, 1, weight=3)  
    G.add_edge(0, 3, weight=7)  
    G.add_edge(1, 2, weight=1)  
    G.add_edge(1, 4, weight=8) 
    G.add_edge(2, 3, weight=2)  
    G.add_edge(3, 4, weight=3)  
    G.add_edge(4, 0, weight=4)  
    
    print(f"Número de nodos: {G.number_of_nodes()}")
    print(f"Número de aristas: {G.number_of_edges()}")
    print("\nNodos:")
    for node in G.nodes(data=True):
        print(f"  {node[0]}: {node[1]['name']}")
    
    print("\nAristas:")
    for u, v, data in G.edges(data=True):
        print(f"  {G.nodes[u]['name']} -> {G.nodes[v]['name']}: {data['weight']}")
    
    distance_matrix = nx.floyd_warshall(G, weight='weight')

    nodes = list(G.nodes())
    n = len(nodes)
    result_matrix = np.zeros((n, n))
    
    for i in range(n):
        for j in range(n):
            if distance_matrix[i][j] == float('inf'):
                result_matrix[i][j] = float('inf')
            else:
                result_matrix[i][j] = distance_matrix[i][j]
    
    print("\nMatriz de distancias mínimas (resultado de Floyd-Warshall):")
    for i in range(n):
        for j in range(n):
            if result_matrix[i][j] == float('inf'):
                print("∞\t", end="")
            else:
                print(f"{result_matrix[i][j]:.0f}\t", end="")
        print()
    
    eccentricity = {}
    for node in G.nodes():
        max_distance = 0
        for target in G.nodes():
            if node != target and distance_matrix[node][target] != float('inf'):
                max_distance = max(max_distance, distance_matrix[node][target])
        eccentricity[node] = max_distance
    
    center = min(eccentricity.items(), key=lambda x: x[1])[0]
    print(f"\nEl centro del grafo es: {G.nodes[center]['name']}")
    
    print("\nRutas más eficientes desde Ciudad de Guatemala:")
    for target in range(n):
        if target != 0:  
            if distance_matrix[0][target] == float('inf'):
                print(f"Hacia {G.nodes[target]['name']}: No hay ruta disponible")
            else:
                print(f"Hacia {G.nodes[target]['name']}: {distance_matrix[0][target]:.0f} km")

if __name__ == "__main__":
    main()